sdk/include/ds/linked_list.h

```diff
@@ -0,0 +1,642 @@
+	 * exclusively to `to`, applying `f` to each cons cell in turn.  If `f`
+	 * returns `true`, the search stops early and returns `true`; otherwise,
+	 * search returns `false`.  To (side-effectfully) visit every node in the
+	 * span, have `f` always return false.
+	 */
+	template<cell::HasPrimOps Cell, typename F>
+	__always_inline bool search(Cell *from, Cell *to, F f)
+	{
+		Cell *elem;
+		for (elem = from; elem != to; elem = elem->cell_next())
+		{
+			if (f(elem))
+			{
+				return true;
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Search through all elements of a ring *except* `elem`.  If `elem` is the
+	 * sentinel of a ring, then this is, as one expects, a `search` over all
+	 * non-sentinel memebers of the ring.
+	 */
+	template<cell::HasPrimOps Cell, typename F>
+	__always_inline bool search(Cell *elem, F f)
+	{
+		return search(static_cast<Cell *>(elem->cell_next()), elem, f);
+	}
+
+	/**
+	 * Convenience wrapper for a sentinel cons cell, encapsulating some common
+	 * patterns.
+	 */
+	template<cell::HasPrimOpsInit Cell_>
+	struct Sentinel
+	{
+		using Cell = Cell_;
+
+		/**
+		 * The sentinel node itself.  Viewing the ring as a list, this
+		 * effectively serves as pointers to the head (next) and to the tail
+		 * (prev) of the list.  Unlike more traditional nullptr-terminated
+		 * lists, though, here, the sentinel participates in the ring.
+		 *
+		 * This is marked `cheri_no_subobject_bounds` because some of our cons
+		 * cell implementations use pointer proxies that rely on the bounds
+		 * provided by `this` (which, in turn, is likely to be
+		 * `cheri_no_subobject_bounds`)
+		 */
+		Cell sentinel __attribute__((__cheri_no_subobject_bounds__));
+
+		__always_inline void init()
+		{
+			sentinel.cell_init();
+		}
+
+		__always_inline bool is_empty()
+		{
+			return linked_list::is_singleton(&sentinel);
+		}
+
+		__always_inline void append(Cell *elem)
+		{
+			linked_list::insert_before(&sentinel, elem);
+		}
+
+		__always_inline void append_new(Cell *elem)
+		{
+			linked_list::insert_new_before(&sentinel, elem);
+		}
+
+		__always_inline void prepend(Cell *elem)
+		{
+			linked_list::insert_before(elem, &sentinel);
+		}
+
+		__always_inline Cell *first()
+		{
+			return sentinel.cell_next();
+		}
+
+		__always_inline Cell *last()
+		{
+			return sentinel.cell_prev();
+		}
+
+		__always_inline bool last_is_eq(Cell *p)
+		{
+			return p->cell_next_is_eq(&sentinel);
+		}
+
+		__always_inline Cell *unsafe_take_first()
+		{
+			Cell *f = sentinel.cell_next();
+			linked_list::unsafe_remove_link(&sentinel, f);
+			return f;
+		}
+
+		__always_inline Cell *take_all()
+		{
+			auto p = linked_list::unsafe_remove(&sentinel);
+			sentinel.cell_init();
+			return p;
+		}
+
+		template<typename F>
+		__always_inline bool search(F f)
+		{
+			return linked_list::search(&sentinel, f);
+		}
+	};
+
+	namespace cell
+	{
+
+		/** Cons cell using two pointers */
+		class Pointer
+		{
+			Pointer *prev, *next;
```

Comment:
```suggestion
			Pointer *prev = this;
                        Pointer *next = this;
```

Reply:
I'm open to being persuaded that cons cells should have constructors (hah), but...
- if we're going to here, we probably should for all three types defined in this file as well as `MChunkHeader` over in `alloc.h`, which we treat as a cons cell with strange pointer proxies.
- we tend not to construct our cells (even using placement new) so much as coerce other pointers into cell types (usually implicitly by coercing a pointer into a cell-containing type like `MChunk`; those, too, do not get constructed in general) which we then `emplace` into their linked lists, so I'm not sure what this buys us.

Reply:
This is skirting dangerously close to UB and probably crossing the line.  If you are changing the type of memory, you must do a placement `new`.  I think the compiler may do some exciting things if you just coerce (in particular, assuming a pointer doesn't alias itself).